controller/api/destination/watcher/opaque_ports_watcher.go

```diff
@@ -0,0 +1,232 @@
+	}
+	k8sAPI.Svc().Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    opw.addService,
+		DeleteFunc: opw.deleteService,
+		UpdateFunc: func(_, obj interface{}) { opw.addService(obj) },
+	})
+	return opw
+}
+
+// Subscribe subscribes a listener to a service; each time the service
+// changes, the listener will be updated if the list of opaque ports
+// changes.
+func (opw *OpaquePortsWatcher) Subscribe(id ServiceID, listener OpaquePortsUpdateListener) error {
+	opw.Lock()
+	defer opw.Unlock()
+	svc, _ := opw.k8sAPI.Svc().Lister().Services(id.Namespace).Get(id.Name)
+	if svc != nil && svc.Spec.Type == corev1.ServiceTypeExternalName {
+		return invalidService(id.String())
+	}
+	opw.log.Infof("Starting watch on service %s", id)
+	ss, ok := opw.subscriptions[id]
+	// If there is no watched service, create a subscription for the service
+	// and no opaque ports
+	if !ok {
+		opw.subscriptions[id] = &svcSubscriptions{
+			opaquePorts: make(map[uint32]struct{}),
+			listeners:   []OpaquePortsUpdateListener{listener},
+		}
+		return nil
+	}
+	// There are subscriptions for this service, so add the listener to the
+	// service listeners. If there are opaque ports for the service, update
+	// the listener with that value.
+	ss.listeners = append(ss.listeners, listener)
+	if len(ss.opaquePorts) != 0 {
+		listener.UpdateService(ss.opaquePorts)
+	}
+	return nil
+}
+
+// Unsubscribe unsubscries a listener from service.
+func (opw *OpaquePortsWatcher) Unsubscribe(id ServiceID, listener OpaquePortsUpdateListener) {
+	opw.Lock()
+	defer opw.Unlock()
+	opw.log.Infof("Stopping watch on service %s", id)
+	ss, ok := opw.subscriptions[id]
+	if !ok {
+		opw.log.Errorf("Cannot unsubscribe from unknown service %s", id)
+		return
+	}
+	for i, l := range ss.listeners {
+		if l == listener {
+			n := len(ss.listeners)
+			ss.listeners[i] = ss.listeners[n-1]
+			ss.listeners[n-1] = nil
+			ss.listeners = ss.listeners[:n-1]
+		}
+	}
+}
+
+func (opw *OpaquePortsWatcher) addService(obj interface{}) {
+	opw.Lock()
+	defer opw.Unlock()
+	svc := obj.(*corev1.Service)
+	if svc.Namespace == kubeSystem {
+		return
+	}
+	id := ServiceID{
+		Namespace: svc.Namespace,
+		Name:      svc.Name,
+	}
+	opaquePorts, err := getServiceOpaquePortsAnnotation(svc)
+	if err != nil {
+		opw.log.Errorf("failed to get %s service opaque ports annotation: %s", id, err)
+		return
+	}
+	// If the service has no opaque ports there are no updates to send.
+	if len(opaquePorts) == 0 {
+		return
+	}
+	ss, ok := opw.subscriptions[id]
+	// If there are no subscriptions for this service, create one with the
+	// opaque ports.
+	if !ok {
+		opw.subscriptions[id] = &svcSubscriptions{
+			opaquePorts: opaquePorts,
+			listeners:   []OpaquePortsUpdateListener{},
+		}
+		return
+	}
+	// Do not send updates if there was no change in the opaque ports; if
+	// there was, send an update to each listener.
+	if !reflect.DeepEqual(ss.opaquePorts, opaquePorts) {
```

Comment:
I like to avoid reflect if at all possible.  You can write a quick function for comparing two maps by iterating through them.  It's the "go way".

Reply:
Sounds good I removed it. I thought we used it in a lot more places based off searching the uses of `reflect.DeepEqual` in the code, but looks like a lot is pointer comparisons. Not sure of other places we use this for maps.